src/Compilers/CSharp/Portable/BoundTree/BoundTreeRewriter.cs

```diff
@@ -59,4 +60,94 @@ public virtual TypeSymbol VisitType(TypeSymbol type)
             return list;
         }
     }
+
+    internal abstract class BoundTreeRewriterWithStackGuard : BoundTreeRewriter
+    {
+        private int _recursionDepth;
+
+        protected BoundTreeRewriterWithStackGuard()
+        { }
+
+        protected BoundTreeRewriterWithStackGuard(int recursionDepth)
+        {
+            _recursionDepth = recursionDepth;
+        }
+
+        protected int RecursionDepth => _recursionDepth;
+
+        public override BoundNode Visit(BoundNode node)
+        {
+            var expression = node as BoundExpression;
+            if (expression != null)
+            {
+                return VisitExpressionWithStackGuard(ref _recursionDepth, expression);
+            }
+
+            return base.Visit(node);
+        }
+
+        protected BoundExpression VisitExpressionWithStackGuard(BoundExpression node)
+        {
+            return VisitExpressionWithStackGuard(ref _recursionDepth, node);
+        }
+
+        protected sealed override BoundExpression VisitExpressionWithoutStackGuard(BoundExpression node)
+        {
+            return (BoundExpression)base.Visit(node);
+        }
+    }
+
+    internal abstract class BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator : BoundTreeRewriterWithStackGuard
+    {
+        protected BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator()
+        { }
+
+        protected BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator(int recursionDepth)
+            : base(recursionDepth)
+        { }
+
+        public sealed override BoundNode VisitBinaryOperator(BoundBinaryOperator node)
+        {
+            BoundExpression child = node.Left;
+
+            if (child.Kind != BoundKind.BinaryOperator)
+            {
+                return base.VisitBinaryOperator(node);
```

Comment:
Is this an attempt to avoid allocation a stack if we're not nested in an expression? I notice that this code is almost identical to code below.